src/proto/media/file.rs

```diff
@@ -226,3 +433,385 @@ bitflags! {
         const VALID_ATTR = 0x37;
     }
 }
+
+/// Common trait for data structures that can be used with
+/// File::set_info() or File::set_info().
+///
+/// The long-winded name is needed because "FileInfo" is already taken by UEFI.
+pub trait FileProtocolInfo: Identify {
+    /// Turn an UEFI-provided pointer-to-base into a Rust-style fat reference
+    unsafe fn from_uefi<'a>(ptr: *mut c_void) -> &'a mut Self;
+}
+
+/// Dynamically sized FileProtocolInfo with a header and an UCS-2 name
+///
+/// All struct that can currently be queried via Get/SetInfo can be described as
+/// a (possibly empty) header followed by a variable-sized name.
+///
+/// Since such dynamic-sized types are a bit unpleasant to handle in Rust today,
+/// this generic struct was created to deduplicate the relevant code.
+///
+/// The reason why this struct covers the whole DST, as opposed to the [Char16]
+/// part only, is that pointers to DSTs are created in a rather unintuitive way
+/// that is best kept centralized in one place.
+#[repr(C)]
+pub struct NamedFileProtocolInfo<Header: FileProtocolInfoHeader> {
+    header: Header,
+    name: [Char16],
+}
+
+/// Common properties of headers that can be used inside NamedFileProtocolInfo
+///
+/// The safety of NamedFileProtocolInfo relies on the following conditions:
+///
+/// - The GUID matches an info structure definition in the UEFI File Protocol
+/// - The header type's data layout matches the UEFI specification
+pub unsafe trait FileProtocolInfoHeader {
+    /// GUID of the full NamedFileProtocolInfo
+    const GUID: Guid;
```

Comment:
That would not feel right to me because what is being identified is the full NamedFileProtocolInfo<Header>, not the header. However, it may actually be a good idea to rework the trait infrastructure here so that the people directly implement Identify + FileProtocolInfo on the full type, rather than implementing a custom trait on the header and having NamedFileProtocolInfo impl the rest.

Reply:
Fixed